descriptors/ModuleDescriptor-template.json

```
@@ -18,7 +18,7 @@
         {
           "methods": [ "POST" ],
           "pathPattern": "/bulk-operations/{operationId}/content-update",
-          "permissionsRequired": [ "bulk-operations.content-update.collection.post" ],
+          "permissionsRequired": [ "bulk-operations.content-update.item.post" ],
```

If it were me, I might be tempted to switch the ordering of content-update and collection / item.

I'm suggesting this because I think of the content update being an operation on an item, thus it reads more intuitively to me.

I can understand not wanting to do this because it might mean changing a bunch of other permission names too, in order to fit this form. For example, this would affect the preview and start operations etc as well.

```suggestion
          "permissionsRequired": [ "bulk-operations.item.content-update.post" ],
```

---

Similarly, the `bulk-operations.errors.collection.get` permission name also uses collection when I think it should use item.

---

`bulk-operations.errors.collection.get` returns a collection of errors, https://github.com/folio-org/mod-bulk-operations/blob/master/src/main/resources/swagger.api/schemas/errors.json

so I believe it's applicable to use it here

---

@NikitaSedyx Thank you for your comment. 

I think this might be a misinterpretation of the naming conventions for permissions.

The item vs. collection part is not about the content submitted to the endpoint or the response it produces, it is about the context of the endpoint relative to other endpoints.

This is quite often not obvious, or even helpful, in endpoints that aren’t part of a collection API, and thus I would tend to omit them.

        {
          "methods": [ "POST" ],
          "pathPattern": "/bulk-operations/{operationId}/content-update",
          "permissionsRequired": [ "bulk-operations.content-update.collection.post" ],
          "modulePermissions": [
          ]
        }
In the context of this example, I would consider this an item level endpoint because it is in the context of a single bulk operation (the presence of the {operationId} path parameter tells me this).

I don’t think it matters that the endpoint itself receives or responds with a collection (presumably modelled as an array) of rules and contents because they are related to a single bulk operation.

As another example, I wouldn’t consider updating a single instance record (that is part of the collection) to be a collection operation because the JSON representation it receives contains collection properties e.g. what subjects the work relates to.

Does that make sense?

---

@marcjohnson-kint for `content-update` definitely yes, but `errors` should be a collection IMO

in fact our naming convention is more applicable for REST resources, and it's not so easy to apply it for smth non very standard

---

@NikitaSedyx 

> for `content-update` definitely yes, but `errors` should be a collection IMO

As I understand it, both content-update and errors are both relative to a single bulk operation.

What distinction about them have I missed that means one of them is considered related to a single item and the other is related to the whole collection?

```
{
    "methods" : [ "GET" ],
    "pathPattern" : "/bulk-operations/{operationId}/errors",
    "permissionsRequired" : [
        "bulk-operations.errors.collection.get"
    ],
    "modulePermissions" : [
        "bulk-edit.errors.collection.get"
    ]
}
```

> in fact our naming convention is more applicable for REST resources, and it's not so easy to apply it for smth non very standard

I agree, this convention is mostly designed for CRUD style* HTTP APIs and is not easy to apply to or does not fit into other styles of HTTP API.

( * I prefer not to call them ReST because my interpretation of ReST is somewhat different to that style)

---

> > As I understand it, both content-update and errors are both relative to a single bulk operation.
> 
> from this perspective it's true, so following this permissions name should be like `bulk-operations.item.errors.collection.get` 

Sure, I think that would be ok.

I think the second classification (`collection`) is unnecessary because the `item` or `collection` distinction is primarily to distinguish with endpoints are operating on a collection of records or an individual record. 

As I understand it, there is no way to interact with individual errors (which aren't records in their own right), only the whole collection of them.

> and might interpreted as `get errors collection for bulk operation`

That sounds like a reasonable description of what the endpoint does, so that fits for me.